### PR TITLE
Bump llvm 2022-11-22

### PR DESCRIPTION
**Pull request by richardxia**
Leaving this as a draft since this is a WIP.

@mortbopet, I am encountering an issue with the bump where the SCFToCalyx converter will loop forever. I've done a bit of investigating, and I think I understand the problem, but I believe it's going to require a pretty substantial change to the SCFToCalyx converter's structure.

I git bisected the LLVM changes and identified https://github.com/llvm/llvm-project/commit/2fea658a74196a9c9128be34c5bc306eba7a025e as the commit that causes the SCFToCalyx converter to loop indefinitely. In that commit, the `GreedyPatternRewriter` was changed such that when `updateRootInPlace()` is called on an op, at the very end it will push that op back onto the worklist. The rationale is that the pattern rewriter will modify that root op, and as long as the pattern continues to match the root op, it implies that the root op can be further reduced by executing the rewriter again. Note that the upstream commit message also says:

> Note: If your project goes into an infinite loop because of this change, you likely have one or multiple faulty patterns that modify the same operations in-place (`updateRootInplace`) indefinitely.

I wasn't previously familiar with the SCFToCalyx converter, but after having read through the code, I think the issue is that [FuncOpConversion](https://github.com/llvm/circt/blob/e1bd9e1161a47cadfe3bd26cdc918f2fc3940bc9/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L713), which is a RewritePattern, operates on a FuncOp, but instead of modifying the FuncOp, it inserts a ComponentOp after the FuncOp. FuncOpConversion doesn't appear to ever return `failure()`, which means that according to the semantics of a RewritePattern, the pattern "always matches," even if we previously ran the RewritePattern on the FuncOp.

Prior to the upstream MLIR change to `GreedyPatternRewriter`, this didn't cause issues because the FuncOp wasn't getting pushed back onto the worklist, so it would only ever run once. However, after the upstream change, the FuncOp is getting pushed back onto the worklist, and since the pattern "matches," we keep running it over and over again, pushing the FuncOp back into the worklist over and over again (and probably inserting many duplicate ComponentOps in the process).

I've tried thinking of different solutions to this problem, but I think fundamentally we are just misusing RewritePatterns if we are not actually using them to rewrite an Op. I'm not too familiar with the actual conversion logic in the SCFToCalyx file, but I'm wondering if it would be better to just not use the PatternRewriter infrastructure and instead just build up the Calyx ops in a more direct way.

If you have the time, @mortbopet, could you take a look at this and make changes to the SCFToCalyx code such that they work on the latest LLVM tree? Note that this Thursday and Friday are a holiday in the US, so I as well as many of the others in the US probably won't be as available in the next few days, but I would appreciate it if you could take a look.

cc @mikeurbach

### CODE REVIEW

**Comment by richardxia** [comment]
@mortbopet, I'm not sure if you saw this PR, but I'm blocked on bumping the llvm submodule in circt due to issues with the SCFToCalyx conversion. I think the right fix will involve substantial changes to SCFToCalyx, so I wanted to get your thoughts on the issue describe above.

**Comment by mikeurbach** [comment]
I'm going to take a quick look at this, since I have some familiarity with this pass.

**Comment by mortbopet** [comment]
@richardxia sorry, just noticed now. I'll have a review ready sometime tomorrow - thanks for working on a fix for this.

**Comment by mikeurbach** [comment]
I tried a few things, none of which were successful:
* update [PartialLoweringPattern](https://github.com/llvm/circt/blob/0f47c4f5ba7b384369330dbec262652b6e593fb7/include/circt/Dialect/Calyx/CalyxLoweringUtils.h#L463) to have some state and early exit from matchAndRewrite if it has already run. this doesn't work because matchAndRewrite is supposed to be const, and patterns aren't supposed to have this kind of state. maybe there is a way to hack this in
* update [FuncOpConversion](https://github.com/llvm/circt/blob/0f47c4f5ba7b384369330dbec262652b6e593fb7/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L717) to return early if the component op is already present in `functionMapping`. this doesn't break the loop, since the parent `PartialLoweringPattern` is still going to call `updateRootInPlace`, which will push the FuncOp back onto the greedy worklist regardless.
* updated [runPartialPattern](https://github.com/llvm/circt/blob/0f47c4f5ba7b384369330dbec262652b6e593fb7/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L1253) to use `applyOpPatternsAndFold` on `getOperation()` instead of `applyPatternsAndFoldGreedily`. this didn't loop forever, but it also didn't do anything, since the patterns don't target the op returned by `getOperation()`. maybe this driver would work if we found the right operation to run the pattern on, or tried it on all children ops of `getOperation()` or something like that.
* updated [runPartialPattern](https://github.com/llvm/circt/blob/0f47c4f5ba7b384369330dbec262652b6e593fb7/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L1253) to use `applyOpPatternsAndFold` on `getOperation().getOps()`. this driver also looped forever.

The root issue here is a) `updateRootInPlace` is always called by `PartialLoweringPattern::matchAndRewrite`, which will now always push the op back onto the worklist, and b) the patterns are not idempotent. I think dealing with one or the other should fix this issue.

**Comment by mikeurbach** [comment]
> or tried it on all children ops of getOperation() or something like that.

I tried this as well and now this driver does something, but then crashes in an assert here in `ComponentLoweringStateInterface::addReturnReg`. Maybe this is a way forward. Here's what I did: https://github.com/llvm/circt/compare/mikeurbach/llvm-bump-11-2022?expand=1

**Comment by mortbopet** [comment]
> update [PartialLoweringPattern](https://github.com/llvm/circt/blob/0f47c4f5ba7b384369330dbec262652b6e593fb7/include/circt/Dialect/Calyx/CalyxLoweringUtils.h#L463) to have some state and early exit from matchAndRewrite if it has already run. this doesn't work because matchAndRewrite is supposed to be const, and patterns aren't supposed to have this kind of state. maybe there is a way to hack this in.

A hack of this kind has been used in the struct flattening pass by having [a mutable variable in the pattern to communicate whether the pattern was applied[(https://github.com/llvm/circt/blob/main/lib/Dialect/HW/Transforms/FlattenIO.cpp#L72) which is then used as a [legality condition for a given op](https://github.com/llvm/circt/blob/main/lib/Dialect/HW/Transforms/FlattenIO.cpp#L265-L266).

**Comment by mikeurbach** [comment]
Alright I tried something similar to what @mortbopet suggested in [2466431](https://github.com/llvm/circt/pull/4353/commits/2466431f2e2be370563cb1cafb410fc5401b9609), and I believe the tests pass now. At least they did locally.

**Comment by richardxia** [comment]
Awesome, thanks for pushing up that commit! I can confirm that it works locally for me. I'd hit the Approve button on the PR, but I can't because I was the one who opened the PR in the first place.

**Review by mikeurbach — APPROVED** [review]
Wellp, I guess I will approve it, even though it was my change. Should be good to go once CI passes.

**Comment by richardxia** [comment]
Thanks! Since the CI checks passed, I'm going to hit merge on this. Thanks again for making that last commit to get the SCFToCalyx conversion working!

**Pull request merged**
